[DOC-11643] New SQL++ Statements for 8.0

rakhi-prathap opened the pull request:
Jira: DOC-11643

This PR adds the following new SQL++ statements.

- Users - [CREATE USER](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/createuser.html), [ALTER USER](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/alteruser.html), [DROP USER](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/dropuser.html)
- Groups - [CREATE GROUP](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/creategroup.html), [ALTER GROUP](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/altergroup.html), [DROP GROUP](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/dropgroup.html)
- Buckets - [CREATE BUCKET](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/createbucket.html), [ALTER BUCKET](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/alterbucket.html),[ DROP BUCKET](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/dropbucket.html)

It also updates two existing statements to include the keywords `user/users` and `group/groups`. 

- [GRANT](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/grant.html)
- [REVOKE](https://preview.docs-test.couchbase.com/docs-devex-DOC-11643-new-statements/server/current/n1ql/n1ql-language-reference/revoke.html)

rakhi-prathap committed:
grant and revoke stmts

rakhi-prathap committed:
Add CREATE USER

rakhi-prathap committed:
Add ALTER USER

rakhi-prathap committed:
Add DROP

rakhi-prathap committed:
Add CREATE GROUP

rakhi-prathap committed:
Update CREATE GROUP

rakhi-prathap committed:
Add DROP GROUP

rakhi-prathap committed:
Update GRANT

rakhi-prathap committed:
Update REVOKE

rakhi-prathap committed:
More statements

rakhi-prathap committed:
Minor edits to GRANT and REVOKE

rakhi-prathap committed:
Adding new SQL++ statements

rakhi-prathap committed:
Update GRANT and REVOKE statements

rakhi-prathap committed:
Refine the content

rakhi-prathap committed:
Update HEAD.yml

rakhi-prathap committed:
Add a custom preview.yml

rakhi-prathap committed:
Revert preview.yml changes

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/createuser.adoc:
```diff
@@ -0,0 +1,83 @@
+
+[horizontal]
+username::
+(Required) The unique identifier for the new user. 
```
The username is the local user's identifier. It is important to mention the word "local" as local users and external users are different. 

Check: https://docs.couchbase.com/server/current/learn/security/authentication-domains.html

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/createuser.adoc:
```diff
@@ -0,0 +1,83 @@
+
+== Syntax
+
+[source,ebnf]
```
There was a recent change that needs to be added to the documentation:

An optional IF NOT EXISTS clause must be added.

The grammar is now:

'CREATE' 'USER' ( 'IF' 'NOT' 'EXISTS' )? username 'PASSWORD' '"' password '"' ('WITH' '"' name '"')? ('GROUP' group | 'GROUPS' group (',' group)* | 'NO' 'GROUPS')?

IF NOT EXISTS clause:
The optional IF NOT EXISTS clause enables the statement to complete successfully when the specified user already exists. If a user with the same username already exists, then:

If this clause is not present, an error is generated.

If this clause is present, the statement does nothing and completes without error.

Check: https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/createcollection.html#if-not-exists 

as reference for the clause

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/alteruser.adoc:
```diff
@@ -0,0 +1,75 @@
+
+[horizontal]
+username::
+(Required) The unique identifier of the user you want to update.
```
The username is the local user's identifier. It is important to mention the word "local" as local users and external users are different.

Check: https://docs.couchbase.com/server/current/learn/security/authentication-domains.html

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/dropuser.adoc:
```diff
@@ -0,0 +1,35 @@
+image::n1ql-language-reference/drop-user.png["Syntax diagram: refer to source code listing", align=left]
+
+[horizontal]
+username::
```
The username is the local user's identifier. It is important to mention the word "local" as local users and external users are different.

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/dropuser.adoc:
```diff
@@ -0,0 +1,35 @@
+
+== Syntax
+
+[source,ebnf]
```
There was a recent change to add the optional IF EXISTS clause to the DROP USER statement. This needs to be added to the documentation:

The grammar should be:

`'DROP' 'USER' ( 'IF' 'EXISTS' )? username`

The optional IF EXISTS clause enables the statement to complete successfully when the specified user doesn’t exist. If the username does not exist, then:

If this clause is not present, an error is generated.

If this clause is present, the statement does nothing and completes without error.


Check this for reference on the IF EXISTS clause - https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/dropcollection.html#if-exists

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/creategroup.adoc:
```diff
@@ -0,0 +1,155 @@
+
+== Syntax
+
+[source,ebnf]
```
There was a recent change that needs to be added to the documentation:

An optional IF NOT EXISTS clause must be added.

The grammar is now:

`'CREATE' 'GROUP' ( 'IF' 'NOT' 'EXISTS' )? name ('WITH' '"' description '"')? ('ROLE' role | 'ROLES' role (',' role)* | NO ROLES )`
 
IF NOT EXISTS clause:

The optional IF NOT EXISTS clause enables the statement to complete successfully when the specified group already exists. If a group with the same name already exists, then:

If this clause is not present, an error is generated.

If this clause is present, the statement does nothing and completes without error.

Check: https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/createcollection.html#if-not-exists

as reference for the clause

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/creategroup.adoc:
```diff
@@ -0,0 +1,155 @@
+====
+[source,sqlpp]
+----
+CREATE GROUP sales ROLE query_select;
```
This statement is syntactically incorrect. You need to specify a target keyspace to assign query_select privileges on.

For example:

`CREATE GROUP sales ROLE query_select ON default:bucket1.scope1.collection1;`

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/creategroup.adoc:
```diff
@@ -0,0 +1,155 @@
+* `query_update` → `update`
+* `query_delete` → `delete`
+
+target-keyspace::
```
The syntax of target-keyspace is incorrect for this statement.

There should be no optional AS clause.

The syntax for keyspace-ref should be used instead as the syntax for target-keyspace.

Please remove all references to the "AS alias" and modify the sytax for target-keyspace.

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/altergroup.adoc:
```diff
@@ -0,0 +1,164 @@
+* `query_update` → `update`
+* `query_delete` → `delete`
+
+target-keyspace::
```
The syntax of target-keyspace is incorrect for this statement.

There should be no optional AS clause.

The syntax for keyspace-ref should be used instead as the syntax for target-keyspace.

Please remove all references to the "AS alias" and modify the sytax for target-keyspace.

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/dropgroup.adoc:
```diff
@@ -0,0 +1,41 @@
+
+[source,ebnf]
+----
+include::partial$grammar/ddl.ebnf[tag=drop-group]
```
There was a recent change to add the optional IF EXISTS clause to the DROP GROUP statement. This needs to be added to the documentation:

The grammar should be:

`'DROP' 'GROUP' ('IF' 'EXISTS' )? groupname`

The optional IF EXISTS clause enables the statement to complete successfully when the specified group doesn’t exist. If a group with the same does not exist, then:

If this clause is not present, an error is generated.

If this clause is present, the statement does nothing and completes without error.

Check this for reference on the IF EXISTS clause - https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/dropcollection.html#if-exists

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/createbucket.adoc:
```diff
@@ -0,0 +1,81 @@
+(Optional) A JSON object containing a list of name-value pairs that specify additional options for the bucket.
+For a list of valid fields names and values, see {bucket-parameters}[Bucket Parameter Groups] in the REST API documentation.
+
+NOTE: If `with-fields` is not specified, the bucket uses a default memory quota (`ramQuota`) of 100 MiB. 
```
This behaviour was modified recently. Please change the documentation accordingly:

In with-fields, if "ramQuota" is set, the bucket is configured with this value as the ramQuota.

If it has not been set:
- In with-fields, if "storageBackend" option is  set to "magma" and the "numVBuckets" option is set to 1024, then the ramQuota for the bucket is set to 1024 MiB.
- Otherwise, the bucket is created with a default ramQuota of 100 MiB.

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/createbucket.adoc:
```diff
@@ -0,0 +1,81 @@
+It must be unique within the cluster and cannot be longer than 100 characters. 
+Acceptable characters are A-Z, a-z, 0-9, and the special characters underscore, period, dash, and percent.
+
+with-fields:: 
```
Please add the following information to the section:

Unless you explicitly set the optional bucket settings in `with-fields` (as described in https://docs.couchbase.com/server/current/rest-api/rest-bucket-create.html
), the bucket will be created using the default values for those settings.

For optional parameters/options (specified here: https://docs.couchbase.com/server/current/rest-api/rest-bucket-create.html)  that are not set in `with-fields`, the bucket is configured with default values for those settings.

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/createbucket.adoc:
```diff
@@ -0,0 +1,81 @@
+
+== Syntax
+
+[source,ebnf]
```
There was a recent change that needs to be added to the documentation:

An optional IF NOT EXISTS clause must be added.

The grammar is now:

'CREATE ('BUCKET' | 'DATABASE') ( 'IF' 'NOT' 'EXISTS' )? name ('WITH' '{' with-fields '}')?

IF NOT EXISTS clause:

The optional IF NOT EXISTS clause enables the statement to complete successfully when the specified bucket already exists. If a bucket with the same name already exists, then:

If this clause is not present, an error is generated.

If this clause is present, the statement does nothing and completes without error.

Check: https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/createcollection.html#if-not-exists

as reference for the clause

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/alterbucket.adoc:
```diff
@@ -0,0 +1,73 @@
+= ALTER BUCKET
+:description: The ALTER BUCKET statement enables you to update a bucket.
```
> The ALTER BUCKET statement enables you to update a bucket.


This statement does not update the bucket but rather the bucket's configurations.

So perhaps can we please re-word this to something like...

"...enables you to update a bucket's configurations.."

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/alterbucket.adoc:
```diff
@@ -0,0 +1,73 @@
+
+== Purpose
+
+Use the ALTER BUCKET statement to modify an existing bucket in your Couchbase cluster. 
```
> Use the ALTER BUCKET statement to modify an existing bucket


Modify an existing bucket's configurations**

dhanyagowrish commented on modules/n1ql/pages/n1ql-language-reference/dropbucket.adoc:
```diff
@@ -0,0 +1,60 @@
+
+== Syntax
+
+[source,ebnf]
```
There was a recent change to add the optional IF EXISTS clause to the DROP BUCKET statement. This needs to be added to the documentation:

The grammar should be:

`'DROP' ('BUCKET' | 'DATABASE') ('IF' 'EXISTS' )? name`

The optional IF EXISTS clause enables the statement to complete successfully when the specified bucket doesn’t exist. If a bucket with the same does not exist, then:

If this clause is not present, an error is generated.

If this clause is present, the statement does nothing and completes without error.

Check this for reference on the IF EXISTS clause - https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/dropcollection.html#if-exists

dhanyagowrish commented on modules/n1ql/partials/nav.adoc:
```diff
@@ -101,13 +106,17 @@
+      **** xref:n1ql:n1ql-language-reference/dropbucket.adoc[]
       **** xref:n1ql:n1ql-language-reference/dropcollection.adoc[]
       **** xref:n1ql:n1ql-language-reference/dropfunction.adoc[]
+      **** xref:n1ql:n1ql-language-reference/dropgroup.adoc[]
```
Additionally:

For CREATE USER/ALTER USER/DROP USER and CREATE GROUP/ALTER GROUP/DROP GROUP can we please add the following section:

Prerequisites:

RBAC Privileges:

To execute the statement the user must have either Full Admin or Security Admin role granted.

dhanyagowrish reviewed (commented):
Apologies, there are several changes that were requested, that were never specified in the DOC ticket description.

Thank you for making the changes!

rakhi-prathap committed:
Incorporating tech review feedback

rakhi-prathap committed:
Incorporating tech review feedback

rakhi-prathap committed:
Add new examples using IF EXISTS/IF NOT EXISTS

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -288,3 +309,45 @@ alter-sequence-options ::= ( restart-with
+
+/* tag::alter-group[] */
+alter-group ::= 'ALTER' 'GROUP' name ( 'WITH' '"' description '"' )? 
+                ( 'ROLE' rbac-role | 'ROLES' rbac-role (',' rbac-role )* | NO ROLES )?        
```
Missing quotes, and remove trailing spaces

```suggestion
                ( 'ROLE' rbac-role | 'ROLES' rbac-role (',' rbac-role )* | 'NO' 'ROLES' )?
```

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -288,3 +309,45 @@ alter-sequence-options ::= ( restart-with
+/* tag::create-user[] */
+create-user ::= 'CREATE' 'USER' ( 'IF' 'NOT' 'EXISTS' )? username 'PASSWORD' '"' password '"' 
+                ( 'WITH' '"' name '"' )? 
+                ( 'GROUP' group | 'GROUPS' group ( ',' group )* | 'NO' 'GROUPS' )?                
```
Remove trailing spaces

```suggestion
                ( 'GROUP' group | 'GROUPS' group ( ',' group )* | 'NO' 'GROUPS' )?
```

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -288,3 +309,45 @@ alter-sequence-options ::= ( restart-with
+/* tag::create-group[] */
+create-group ::= 'CREATE' 'GROUP' ( 'IF' 'NOT' 'EXISTS' )? name 
+                 ( 'WITH' '"' description '"' )? 
+                 ( 'ROLE' rbac-role | 'ROLES' rbac-role ( ',' rbac-role )* | 'NO' 'ROLES' )            
```
Remove trailing spaces

```suggestion
                 ( 'ROLE' rbac-role | 'ROLES' rbac-role ( ',' rbac-role )* | 'NO' 'ROLES' )
```

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -288,3 +309,45 @@ alter-sequence-options ::= ( restart-with
+/* end::alter-group[] */
+
+/* tag::rbac-role[] */
+rbac-role ::= role ( 'ON' keyspace-ref )?       
```
Remove trailing spaces

```suggestion
rbac-role ::= role ( 'ON' keyspace-ref )?
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/creategroup.adoc:
```diff
@@ -0,0 +1,147 @@
+====
+[source,sqlpp]
+----
+CREATE GROUP IF NOT EXISTS support ROLE query_update ON `travel-sample`.`inventory`.`airport`;
```
Maybe break this example over two lines

```suggestion
CREATE GROUP IF NOT EXISTS support ROLE query_update
ON `travel-sample`.`inventory`.`airport`;
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/alterbucket.adoc:
```diff
@@ -0,0 +1,75 @@
+== Example
+
+====
+.Alter a bucket and update its memory quota, maximum TTL, and durability level
```
Put the example title before the example delimiter so that the title has the automatic **Example 1.** label

```suggestion
.Alter a bucket and update its memory quota, maximum TTL, and durability level
====
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/alteruser.adoc:
```diff
@@ -0,0 +1,80 @@
+----
+ALTER USER Bob NO GROUPS;
+----
+====
```
Put the example titles before the example delimiters, so that the title has the automatic **Example 1** label

```suggestion
.Change a user's password and full name 
====
[source,sqlpp]
----
ALTER USER Hilary PASSWORD "newpassword" WITH "Hilary Chloe";
----
====

.Assign a user to a new group
====
[source,sqlpp]
----
ALTER USER Alice GROUP support;
----
====

.Remove a user from existing groups
====
[source,sqlpp]
----
ALTER USER Bob NO GROUPS;
----
====
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/createuser.adoc:
```diff
@@ -0,0 +1,102 @@
+----
+CREATE USER IF NOT EXISTS David PASSWORD "davidPass" WITH "David Trantow";
+----
+====
```
Example titles again

```suggestion
.Create a user and specify their full name and password
====
[source,sqlpp]
----
CREATE USER Hilary PASSWORD "password123" WITH "Hilary Smith";
----
====

.Create a user and assign them to a single group
====
[source,sqlpp]
----
CREATE USER Alice PASSWORD "password123" GROUP agents;
----
====

.Create a user and assign them to multiple groups
====
[source,sqlpp]
----
CREATE USER Bob PASSWORD "P@ssw0rd" GROUPS agents, tourguides, support;
----
====

.Create a user with no group assignments
====
[source,sqlpp]
----
CREATE USER Charlie PASSWORD "securePass" NO GROUPS;
----
====

.Create a user if they do not already exist
====
[source,sqlpp]
----
CREATE USER IF NOT EXISTS David PASSWORD "davidPass" WITH "David Trantow";
----
====
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/dropbucket.adoc:
```diff
@@ -0,0 +1,80 @@
+----
+DROP BUCKET IF EXISTS `custom-bucket`;
+----
+====
```
Example titles

```suggestion
.Drop a bucket named `student-records`
====
[source,sqlpp]
----
DROP BUCKET `student-records`;
----
====

.Drop a bucket named `custom-bucket` if it exists
====
[source,sqlpp]
----
DROP BUCKET IF EXISTS `custom-bucket`;
----
====
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/dropgroup.adoc:
```diff
@@ -0,0 +1,61 @@
+----
+DROP GROUP IF EXISTS support;
+----
+====
```
Example titles plus typo

```suggestion
.Delete a group named `sales`
====
[source,sqlpp]
----
DROP GROUP sales;
----
====

.Delete a group named `support` if it exists
====
[source,sqlpp]
----
DROP GROUP IF EXISTS support;
----
====
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/dropuser.adoc:
```diff
@@ -0,0 +1,55 @@
+----
+DROP USER IF EXISTS David;
+----
+====
```
Example titles

```suggestion
.Delete a user named Bob
====
[source,sqlpp]
----
DROP USER Bob;
----
====

.Delete a user named David if they exist
====
[source,sqlpp]
----
DROP USER IF EXISTS David;
----
====
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/dropuser.adoc:
Related links for this page?

rakhi-prathap replied:
Added related links.

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/grant.adoc:
```diff
@@ -36,19 +40,47 @@ include::partial$grammar/dcl.ebnf[tag=grant]
+[NOTE]
+====
+* [Optional] Use the keyword `USER` or `USERS` to grant roles to users. 
+* [Required] Use the keyword `GROUP` or `GROUPS` to grant roles to groups. 
```
Not sure why it says Optional and Required with these bullets?

```suggestion
* Use the keyword `USER` or `USERS` to grant roles to users. 
* Use the keyword `GROUP` or `GROUPS` to grant roles to groups. 
```

rakhi-prathap replied:
Doesn't make sense to me either when I look at it now! 
Changed the phrasing and updated the note.

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/revoke.adoc:
```diff
@@ -37,19 +40,47 @@ include::partial$grammar/dcl.ebnf[tag=revoke]
+[NOTE]
+====
+* [Optional] Use the keyword `USER` or `USERS` to revoke roles from users. 
+* [Required] Use the keyword `GROUP` or `GROUPS` to revoke roles from groups. 
```
Not sure why it says Optional and Required with these bullets?

```suggestion
* Use the keyword `USER` or `USERS` to revoke roles from users. 
* Use the keyword `GROUP` or `GROUPS` to revoke roles from groups. 
```

rakhi-prathap replied:
Changed the phrasing and updated the note.

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/createbucket.adoc:
```diff
@@ -0,0 +1,116 @@
+
+== Related Links
+
+* For an overview buckets, see {buckets}[Buckets].
```
Typo

```suggestion
* For an overview of buckets, see {buckets}[Buckets].
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/alterbucket.adoc:
```diff
@@ -0,0 +1,75 @@
+
+== Related Links
+
+* For an overview buckets, see {buckets}[Buckets].
```
Typo

```suggestion
* For an overview of buckets, see {buckets}[Buckets].
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/dropbucket.adoc:
```diff
@@ -0,0 +1,80 @@
+
+== Related Links
+
+* For an overview buckets, see {buckets}[Buckets].
```
Typo

```suggestion
* For an overview of buckets, see {buckets}[Buckets].
```

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -37,6 +50,14 @@ create-collection ::= 'CREATE' 'COLLECTION' ( ( namespace ':' )? bucket '.' scop
 /* end::create-collection[] */
 
+/* tag::alter-bucket[] */
+alter-bucket ::= 'ALTER' ( 'BUCKET' | 'DATABASE' ) name ( 'WITH' '{' with-fields '}' )?
```
As above. Specify in the text that `with-fields` is an object

```suggestion
alter-bucket ::= 'ALTER' ( 'BUCKET' | 'DATABASE' ) name ( 'WITH' with-fields )?
```

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -288,3 +309,45 @@ alter-sequence-options ::= ( restart-with
+
+/* tag::create-user[] */
+create-user ::= 'CREATE' 'USER' ( 'IF' 'NOT' 'EXISTS' )? username 'PASSWORD' '"' password '"' 
+                ( 'WITH' '"' name '"' )? 
```
Similarly, I'm not sure that you should explicitly specify the quotes around the string. What if the password is an expression (e.g. a named parameter) that resolves to a string? Instead, specify in the text that `password` and `name` must be strings.

```suggestion
create-user ::= 'CREATE' 'USER' ( 'IF' 'NOT' 'EXISTS' )? username 'PASSWORD' password 
                ( 'WITH' name )? 
```

rakhi-prathap replied:
Sorry, I went with what was in the doc ticket thinking that's how it was to be.
Have updated it now.

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -288,3 +309,45 @@ alter-sequence-options ::= ( restart-with
+
+/* tag::alter-user[] */
+alter-user ::= 'ALTER' 'USER' username ( 'PASSWORD' '"' password '"' )? 
+                ( 'WITH' '"' name '"' )? 
```
Not sure about quotes around strings. Specify in the text that it is a string

```suggestion
alter-user ::= 'ALTER' 'USER' username ( 'PASSWORD' password )? 
                ( 'WITH' name )? 
```

rakhi-prathap replied:
Sorry, I went with what was in the doc ticket thinking that's how it was to be.
Have updated it now.

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -288,3 +309,45 @@ alter-sequence-options ::= ( restart-with
+
+/* tag::create-group[] */
+create-group ::= 'CREATE' 'GROUP' ( 'IF' 'NOT' 'EXISTS' )? name 
+                 ( 'WITH' '"' description '"' )? 
```
Not sure about quotes around strings. Specify in the text that it is a string

```suggestion
                 ( 'WITH' description )? 
```

rakhi-prathap replied:
Sorry, I went with what was in the doc ticket thinking that's how it was to be.
Have updated it now.

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -288,3 +309,45 @@ alter-sequence-options ::= ( restart-with
+/* end::create-group[] */
+
+/* tag::alter-group[] */
+alter-group ::= 'ALTER' 'GROUP' name ( 'WITH' '"' description '"' )? 
```
Not sure about quotes around strings.

```suggestion
alter-group ::= 'ALTER' 'GROUP' name ( 'WITH' description )? 
```

rakhi-prathap replied:
Sorry, I went with what was in the doc ticket thinking that's how it was to be.
Have updated it now.

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/alterbucket.adoc:
```diff
@@ -0,0 +1,75 @@
+
+image::n1ql-language-reference/alter-bucket.png["Syntax diagram: refer to source code listing", align=left]
+
+The `BUCKET` and `DATABASE` keywords are synonyms, you can use either of them.
```
I don't like using commas to join separate sentences. Maybe a colon or em dash if you don't think a full stop is justified

```suggestion
The `BUCKET` and `DATABASE` keywords are synonyms: you can use either of them.
```

rakhi-prathap replied:
Updated it to a full stop. I think that's what aligns best with our style guide.

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/createbucket.adoc:
```diff
@@ -0,0 +1,116 @@
+
+image::n1ql-language-reference/create-bucket.png["Syntax diagram: refer to source code listing", align=left]
+
+The `BUCKET` and `DATABASE` keywords are synonyms, you can use either of them. 
```
Comma to join separate sentences

```suggestion
The `BUCKET` and `DATABASE` keywords are synonyms: you can use either of them. 
```

rakhi-prathap replied:
Updated it to a full stop.

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/dropbucket.adoc:
```diff
@@ -0,0 +1,80 @@
+
+image::n1ql-language-reference/drop-bucket.png["Syntax diagram: refer to source code listing", align=left]
+
+The `BUCKET` and `DATABASE` keywords are synonyms, you can use either of them.
```
commas to join separate sentences

```suggestion
The `BUCKET` and `DATABASE` keywords are synonyms: you can use either of them.
```

rakhi-prathap replied:
Updated it to a full stop. I think that's what aligns best with our style guide.

simon-dew reviewed (changes requested):
Sorry there are such a lot of comments. It's a big PR! This is actually very good; all of my comments are small quibbles.

simon-dew commented on modules/n1ql/partials/grammar/ddl.ebnf:
```diff
@@ -10,24 +10,37 @@ create-statement ::= create-scope
 
+/* tag::create-bucket[] */
+create-bucket ::= 'CREATE' ( 'BUCKET' | 'DATABASE' ) ( 'IF' 'NOT' 'EXISTS' )? name 
+                  ( 'WITH' '{' with-fields '}' )?
```
Don't explicitly add the curly brackets before and after the object: it could be an expression that resolves to an object. Specify in the text that `with-fields` is an object

```suggestion
                  ( 'WITH' with-fields )?
```

rakhi-prathap committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/alterbucket.adoc:
```diff
@@ -0,0 +1,75 @@
+= ALTER BUCKET
+:description: The ALTER BUCKET statement enables you to update an existing bucket's configuration.
+:page-topic-type: reference
```
Add Couchbase Server 8.0 flag for long-term users

```suggestion
:page-topic-type: reference
:page-status: Couchbase Server 8.0
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/altergroup.adoc:
```diff
@@ -0,0 +1,141 @@
+= ALTER GROUP
+:description: The ALTER GROUP statement enables you to update an existing group.
+:page-topic-type: reference
```
Add Couchbase Server 8.0 flag for long-term users

```suggestion
:page-topic-type: reference
:page-status: Couchbase Server 8.0
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/alteruser.adoc:
```diff
@@ -0,0 +1,86 @@
+= ALTER USER
+:description: The ALTER USER statement enables you to alter the details of an existing user.
+:page-topic-type: reference
```
Add Couchbase Server 8.0 flag for long-term users

```suggestion
:page-topic-type: reference
:page-status: Couchbase Server 8.0
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/createbucket.adoc:
```diff
@@ -0,0 +1,116 @@
+= CREATE BUCKET
+:description: The CREATE BUCKET statement enables you to create a bucket.
+:page-topic-type: reference
```
Add Couchbase Server 8.0 flag for long-term users

```suggestion
:page-topic-type: reference
:page-status: Couchbase Server 8.0
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/creategroup.adoc:
```diff
@@ -0,0 +1,148 @@
+= CREATE GROUP
+:description: The CREATE GROUP statement enables you to create a group.
+:page-topic-type: reference
```
Add Couchbase Server 8.0 flag for long-term users

```suggestion
:page-topic-type: reference
:page-status: Couchbase Server 8.0
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/createuser.adoc:
```diff
@@ -0,0 +1,113 @@
+= CREATE USER
+:description: The CREATE USER statement enables you to create a user.
+:page-topic-type: reference
```
Add Couchbase Server 8.0 flag for long-term users

```suggestion
:page-topic-type: reference
:page-status: Couchbase Server 8.0
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/dropbucket.adoc:
```diff
@@ -0,0 +1,82 @@
+= DROP BUCKET
+:description: The DROP BUCKET statement enables you to delete a bucket.
+:page-topic-type: reference
```
Add Couchbase Server 8.0 flag for long-term users

```suggestion
:page-topic-type: reference
:page-status: Couchbase Server 8.0
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/dropgroup.adoc:
```diff
@@ -0,0 +1,64 @@
+= DROP GROUP
+:description: The DROP GROUP statement enables you to delete a group.
+:page-topic-type: reference
```
Add Couchbase Server 8.0 flag for long-term users

```suggestion
:page-topic-type: reference
:page-status: Couchbase Server 8.0
```

simon-dew commented on modules/n1ql/pages/n1ql-language-reference/dropuser.adoc:
```diff
@@ -0,0 +1,58 @@
+= DROP USER
+:description: The DROP USER statement enables you to delete a user.
+:page-topic-type: reference
```
Add Couchbase Server 8.0 flag for long-term users

```suggestion
:page-topic-type: reference
:page-status: Couchbase Server 8.0
```

simon-dew commented:
@rakhi-prathap sorry, I just added a new comment to each of the new pages about the `:page-status:` attribute

rakhi-prathap committed:
Apply changes from review

rakhi-prathap committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

rakhi-prathap committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

rakhi-prathap committed:
Minor fix

rakhi-prathap committed:
Apply suggestions from code review

Co-authored-by: Simon Dew <39966290+simon-dew@users.noreply.github.com>

rakhi-prathap committed:
Update EBNF to remove quotes

rakhi-prathap committed:
Resolve conflicts

rakhi-prathap committed:
Add drop-bucket to the drop-statement list

rakhi-prathap commented:
> Sorry there are such a lot of comments. It's a big PR! This is actually very good; all of my comments are small quibbles.

@simon-dew Thanks for the review and sorry about the errors. I've gone through and addressed all your comments.
Do you mind taking another look to see if we're good to merge?

simon-dew reviewed (approved):
LGTM! 👍
